File: languages/js/src/entity/identity/credential_proof.ts
import * as identityEntitiesProto from "../../bridge/proto/identity_entities";

/**
 * Represents the proof associated with a credential, including signature and sparse merkle tree proof.
 */
export class CredentialProof {
  signatureProof: string;
  sparseMtProof?: string;

  /**
   * Constructs an CredentialProof object with the specified parameters.
   * @param signatureProof 
   * @param sparseMtProof 
   */
  constructor(signatureProof: string, sparseMtProof?: string) {
    this.signatureProof = signatureProof;
    this.sparseMtProof = sparseMtProof;
  }

  public toProto(): identityEntitiesProto.CredentialProof {
    return identityEntitiesProto.CredentialProof.fromPartial({
      signatureProof: this.signatureProof,
      sparseMtProof: this.sparseMtProof
    });
  }

  static fromProto(
    r: identityEntitiesProto.CredentialProof
  ): CredentialProof {
    return new CredentialProof(r.signatureProof, r.sparseMtProof);
  }
}

File: languages/go/entity/identity/credential_proof.go
package identity

import "github.com/bloock/bloock-sdk-go/v2/internal/bridge/proto"

// CredentialProof represents the proof associated with a credential, including signature and sparse merkle tree proof.
type CredentialProof struct {
	SignatureProof string
	SparseMtProof  string
}

func NewCredentialProofFromProto(s *proto.CredentialProof) CredentialProof {
	if s == nil {
		return CredentialProof{}
	}
	return CredentialProof{
		SignatureProof: s.GetSignatureProof(),
		SparseMtProof:  s.GetSparseMtProof(),
	}
}

func (c CredentialProof) ToProto() *proto.CredentialProof {
	var sparseMtProof *string

	if c.SparseMtProof == "" {
		sparseMtProof = nil
	} else {
		sparseMtProof = &c.SparseMtProof
	}
	return &proto.CredentialProof{
		SignatureProof: c.SignatureProof,
		SparseMtProof:  sparseMtProof,
	}
}
